### PR TITLE
access: refacto slack plugin + introduce the discord one

### PR DESCRIPTION
**Pull request by hugoShaka**
This PR refactors the Slack plugin so we can add support for other messaging apps more easily, reuse logic, and improve feature parity between plugins. I wanted to add Discord support without duplicating logic and avoid gluing Discord support into the Slack plugin. Refactoring this access-request plugin logic will also allow us to implement [future plugin conventions](https://github.com/gravitational/teleport/pull/16288) only once.

The refacto does 3 things:
- Introduce a `MessagingBot` interface that should contain everything messaging-service specific. The `Bot` structs are becoming a bit more than clients as they also handle the recipient discovery logic (translating from whatever string was passed by the user to a working recipient like a public channel or direct message)
- Standardize `MessageData` and `Recipient` structs. Most access request plugins have the same concepts but named differently, we can use a single model to handle them and leave room for extra custom data if needed (teams will need it).
- Introduce a `BaseApp` struct containing all common teleport logic:
  - initialization/process management
  - connection to Teleport
  - how to treat access requests events
  
I have not checked yet if the abstractions introduced can be easily reused for non-messaging services like ServiceNow and PagerDuty. I think they should.

I would like to get feedback before continuing with those changes. This PR is too big, if you want I can schedule a call to walk through those changes so you have the big picture.

If we're all OK with where this is going my next steps would be:
- cleanup up the TODOs I left in the code
- decide what to do with discord DMs? (currently kinda supported, but you have to speak to the bot first. The alternative will require the bot to become aware of its guild + find a way to identify a user)
- find out what tests to do: add Discord tests? make the existing tests reusable? drop them? use them only on `access/common` 
- have Mattermost use the common base (in an other PR)
- have msteams use the common base (in an other PR)
- add discord release pipeline and helm chart (in 2 other PRs)

### CODE REVIEW

**Comment by Joerger** [comment]
@hugoShaka This is great, these plugins have been in need of generalization for a long time. It will make it so much easier to add new plugins with similar functionality, such as this Discord plugin.

I don't have much feedback for now, it looks solid overall. I suspect that the `config.go` and `main.go` files can also be refactored to be generalized if you get around to it. And there's still quite a few mentions of "Slack" in the common and discord code to cleanup.

It also might be a good idea to separate this into a refactoring PR and a Discord PR based on the first PR, up to you.

**Review by r0mant — COMMENTED** [review]
Overall makes sense for reusing the code between Slack and Discord plugins while keeping them separate executables.

My main ask is let's please not turn this into a giant refactoring of all plugins for the sake of refactoring. It makes sense for Slack/Discord because they're very similar. For others maybe do it very gradually, factoring out common things that can be reused but not trying to fit all of them into the same abstractions.

File: access/slack/slack_test.go [hunk 1]
```diff
@@ -703,6 +707,7 @@ func (s *SlackSuite) TestRace() {
 	<-process.Done()
 	require.NoError(t, raceErr)
 
+	time.Sleep(5 * time.Second)
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply access/slack/slack_test.go:710]: oh, this is a remain of a debug test I made, this can be removed without risk

**Comment by hugoShaka** [comment]
> Overall makes sense for reusing the code between Slack and Discord plugins while keeping them separate executables.
> 
> My main ask is let's please not turn this into a giant refactoring of all plugins for the sake of refactoring. It makes sense for Slack/Discord because they're very similar. For others maybe do it very gradually, factoring out common things that can be reused but not trying to fit all of them into the same abstractions.

Makes sense, those abstractions are designed to fit Mattermost out of the box and MsTeams with little changes as the logic is 95% similar. I don't think they will work with the jira, pagerduty nor email plugins out of the box and I don't plan to look at those 3 plugins now.

**Comment by hugoShaka** [comment]
Thank you for the feedback. I will close this PR and break it in two new PRs: Slack refactoring, and Discord plugin.

This will allow simpler reviews and incremental refactoring of other plugins

**Pull request closed without merging**